mithril-common/src/chain_reader/pallas_chain_reader.rs

```diff
@@ -0,0 +1,349 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use pallas_network::{
+    facades::NodeClient,
+    miniprotocols::chainsync::{BlockContent, NextResponse},
+};
+use pallas_traverse::MultiEraBlock;
+
+use crate::{cardano_block_scanner::ScannedBlock, entities::ChainPoint, CardanoNetwork, StdResult};
+
+use super::{ChainBlockNextAction, ChainBlockReader};
+
+/// [PallasChainReader] reads blocks with 'chainsync' mini-protocol
+pub struct PallasChainReader {
+    socket: PathBuf,
+    network: CardanoNetwork,
+    client: Option<NodeClient>,
+}
+
+impl PallasChainReader {
+    /// Creates a new `PallasChainReader` with the specified socket and network.
+    pub fn new(socket: &Path, network: CardanoNetwork) -> Self {
+        Self {
+            socket: socket.to_owned(),
+            network,
+            client: None,
+        }
+    }
+
+    /// Creates and returns a new `NodeClient` connected to the specified socket.
+    async fn new_client(&self) -> StdResult<NodeClient> {
+        let magic = self.network.code();
+        NodeClient::connect(&self.socket, magic)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainReader failed to create a new client")
+    }
+
+    /// Returns a mutable reference to the client.
+    async fn get_client(&mut self) -> StdResult<&mut NodeClient> {
+        if self.client.is_none() {
+            self.client = Some(self.new_client().await?);
+        }
+
+        self.client
+            .as_mut()
+            .with_context(|| "PallasChainReader failed to get a client")
+    }
+
+    /// Intersects the point of the chain with the given point.
+    async fn intersect_point(&mut self, point: &ChainPoint) -> StdResult<()> {
```

Comment:
Why not naming it `instersection_point` ?

Reply:
Maybe `find_intersect_point` is better as the method does not return the intersection point but just computes it?
@falcucci WDYT?

Reply:
I think that `intersect_point` still okay to me, but `find_intersect_point` is also a good name.

is up to you between these two @jpraynaud

Reply:
Let's use `find_intersect_point` which describes better what's done under the hood :+1:

Reply:
Done